src/app/zap-templates/templates/chip/CHIPClusters-src.zapt

```diff
@@ -2,18 +2,30 @@
 #include "CHIPClusters.h"
 
 #include <app/chip-zcl-zpro-codec-api.h>
+#include <app/im-encoder.h>
 
 namespace chip {
 namespace Controller {
 
+// TODO(@vivien-apple): onCompletion is not used by IM for now.
+// TODO: length should be passed to commands when byte string is in argument list.
+// TODO: Commands should take group id as an argument.
+
 {{#chip_server_clusters}}
 
 // {{asCamelCased name false}} Cluster Commands
 {{#chip_server_cluster_commands}}
 CHIP_ERROR {{asCamelCased clusterName false}}Cluster::{{asCamelCased name false}}(Callback::Callback<> * onCompletion{{#chip_server_cluster_command_arguments}}, {{chipType}} {{asCamelCased label}}{{/chip_server_cluster_command_arguments}})
 {
+#ifdef CHIP_APP_USE_INTERACTION_MODEL
+    (void) onCompletion;
+    // TODO(@vivien-apple): onCompletion is not used by IM for now.
+    chip::app::cluster::{{asCamelCased clusterName false}}::Encode{{asType name}}Command(mDevice->GetCommandSender(), mEndpoint, 0{{#chip_server_cluster_command_arguments}}, {{asCamelCased label}}{{/chip_server_cluster_command_arguments}});
+    return SendCommands();
```

Comment:
> It starts to look better !
> 
> Can we generate the relevant code directly here now ? As I said previously, I would like to get rid of `src/app/encoder.cpp` at some point (when nobody uses it directly anymore), and get the code directly generated here. Since I guess IM encoding is going to replace raw encoding, it is the _new_ `src/encoder.cpp` so it makes sense to me to generate it directly here.
> 
New IM encoder/builder and decoder/parser have been put here, https://github.com/project-chip/connectedhomeip/blob/master/src/app/MessageDef.cpp

IM's command send has been put here
https://github.com/project-chip/connectedhomeip/blob/master/src/app/CommandSender.cpp#L48

Reply:
> Can we generate the relevant code directly here now ? As I said previously, I would like to get rid of src/app/encoder.cpp at some point (when nobody uses it directly anymore), and get the code directly generated here. Since I guess IM encoding is going to replace raw encoding, it is the new src/encoder.cpp so it makes sense to me to generate it directly here.

Probably not, the core issue is that these encoder are not only used by device controller, for example, when processing commands, we also need to encode the returning command (like NetworkProvisioningCluster).

In this case, I think it would be better if we move the cluster related code out of the controller.

Reply:
Right. I guess the issue is that `src/controller/CHIPCluster*` files live into `src/controller` but I think they should be part of `src/app`. In fact I did a similar comment to yours when their first have landed (comment here https://github.com/project-chip/connectedhomeip/pull/3879#discussion_r526073904)

I'm fine to move them out of `src/controller` for my part.